repository/src/main/java/com/epam/esm/repository/impl/UserRepositoryImpl.java

```diff
@@ -0,0 +1,111 @@
+    @PersistenceContext
+    private EntityManager entityManager;
+    private final EntityValidator entityValidator;
+
+    public UserRepositoryImpl(EntityValidator entityValidator) {
+        this.entityValidator = entityValidator;
+    }
+
+    @Override
+    public User save(User entity) {
+        entityValidator.validateEntity(entity);
+
+        entityManager.persist(entity);
+        return entity;
+    }
+
+    @Override
+    public Optional<User> findById(Long id) {
+        return Optional.ofNullable(entityManager.find(User.class, id));
+    }
+
+    @Override
+    public List<User> findAll() {
+        return entityManager.createQuery("SELECT u FROM users u", User.class).getResultList();
+    }
+
+    @Override
+    public List<User> findAllWithPage(int page, int size) throws ValidationException {
+        if (page <= 0 || size <= 0)
+            throw new ValidationException("Page number and page size must be positive");
+
+        CriteriaBuilder cb = entityManager.getCriteriaBuilder();
+        CriteriaQuery<User> query = cb.createQuery(User.class);
+        Root<User> userRoot = query.from(User.class);
+
+        query.select(userRoot);
+
+        TypedQuery<User> typedQuery = entityManager.createQuery(query);
+        typedQuery.setFirstResult((page - 1) * size);
+        typedQuery.setMaxResults(size);
+
+        return typedQuery.getResultList();
+    }
+
+    @Override
+    public void delete(Long id) throws NotFoundException {
+        User user = entityManager.find(User.class, id);
+        if (user == null)
+            throw new NotFoundException("User not found with id: " + id);
+        entityManager.remove(user);
+    }
+
+    @Override
+    public Tag findMostUsedTagOfUserWithHighestOrderCost(Long userId) {
+        CriteriaBuilder cb = entityManager.getCriteriaBuilder();
+        CriteriaQuery<Tuple> criteria = cb.createTupleQuery();
+
+        Root<Order> orderRoot = criteria.from(Order.class);
+        Join<Order, GiftCertificate> giftCertificateJoin = orderRoot.join("giftCertificate");
+        Join<GiftCertificate, Tag> tagJoin = giftCertificateJoin.join("tags");
+
+        criteria.multiselect(
+                tagJoin.get("id").alias("tagId"),
+                tagJoin.get("name").alias("tagName"),
+                cb.sum(orderRoot.get("price")).alias("totalPrice")
+        );
+
+        criteria.where(
+                cb.equal(orderRoot.get("user").get("id"), userId)
+        );
+        criteria.groupBy(tagJoin.get("id"), tagJoin.get("name"));
+        criteria.orderBy(cb.desc(cb.sum(orderRoot.get("price"))));
+
+        TypedQuery<Tuple> query = entityManager.createQuery(criteria);
+        List<Tuple> resultList = query.getResultList();
+
+        if (!resultList.isEmpty()) {
+            Tuple result = resultList.get(0);
+            Long tagId = result.get("tagId", Long.class);
+            String tagName = result.get("tagName", String.class);
+            return new Tag(tagId, tagName);
```

Comment:
Are you sure It works as expected?
Here you are getting the first element from the list. I think this query can not give the required functionality.
Imagine there are these tables:
Tag table
id - 1, name - tech
id - 2, name - book
id -3, name - gym

Gift certificate table
id - 1, price - 1000, tagId - 1
id - 2, price - 2000, tagId - 1
id - 3, price - 3000, tagId - 1
id - 4, price - 3000, tagId - 2
id - 5, price - 2000, tagId - 2
id - 6, price - 1000, tagId - 2
id - 7, price - 4000, tagId - 3

orders
id - 1, cost - 1000, certificateId - 1, userId - 1
id - 2, cost - 2000, certificateId - 2, userId - 1
id - 3, cost - 3000, certificateId - 3, userId - 1
id - 4, cost - 3000, certificateId - 4, userId - 1
id - 5, cost - 2000, certificateId - 5, userId - 1
id - 6, cost - 1000, certificateId - 6, userId - 1
id - 7, cost - 4000, certificateId - 7, userId - 1

User with id 1, bought 7 certificates. The most widely used tags are tech and book, three times. Their prices are the same, 6000. So both of them should be returned.

In addition, get executed native query and demonstrate SQL execution plan for it. https://www.postgresql.org/docs/current/sql-explain.html

Reply:
I thought it should be only one Tag. In task it is said like this I think. But I try to change it.

![image](https://github.com/AkobirToshtemirov/GiftCertificates/assets/88495573/4ceb397e-0137-46d8-a8ac-29bc759a98b7)